File: packages/lynx-http/android/com/sigx/http/HttpEventBus.kt
package com.sigx.http

import org.json.JSONObject
import java.util.UUID
import java.util.concurrent.CopyOnWriteArrayList

/**
 * Process-wide pub/sub bus that owns the conversion of OkHttp callbacks
 * into JS-side event payloads. OkHttp callbacks write here; per-LynxView
 * [HttpPublisher] instances read.
 *
 * Mirrors `WebSocketEventBus` — native lifecycle is decoupled from any
 * specific LynxView so in-flight requests survive view recreation.
 *
 * Payloads are emitted as a **JSON string** whose parsed shape matches the
 * `NativeHttpEvent` interface in `src/types.ts`. A string (rather than a
 * structured `JavaOnlyMap`) survives Lynx 0.5.0's bridge marshalling intact
 * — a map carrying a nested map (the `response` event's `headers`) drops its
 * sibling scalar fields crossing the bridge, so `Response.status` arrived
 * undefined and `res.ok` was false for every request (#342). The JS shim
 * already parses string-form events.
 *
 * Event order per request id: response (once) → progress* → chunk* →
 * done | error.
 */
internal object HttpEventBus {

    private val listeners = CopyOnWriteArrayList<Pair<UUID, (String) -> Unit>>()

    fun addListener(fn: (String) -> Unit): UUID {
        val token = UUID.randomUUID()
        listeners.add(token to fn)
        return token
    }

    fun removeListener(token: UUID) {
        listeners.removeAll { it.first == token }
    }

    fun publishResponse(id: Int, status: Int, statusText: String, headers: Map<String, String>) {
        val obj = JSONObject()
        obj.put("id", id)
        obj.put("type", "response")
        obj.put("status", status)
        obj.put("statusText", statusText)
        val headerObj = JSONObject()
        for ((k, v) in headers) headerObj.put(k, v)
        obj.put("headers", headerObj)
        emit(obj)
    }

    fun publishChunk(id: Int, base64: String) {
        val obj = JSONObject()
        obj.put("id", id)
        obj.put("type", "chunk")
        obj.put("data", base64)
        emit(obj)
    }

    fun publishProgress(id: Int, loaded: Long, total: Long) {
        val obj = JSONObject()
        obj.put("id", id)
        obj.put("type", "progress")
        obj.put("loaded", loaded)
        obj.put("total", total)
        emit(obj)
    }

    fun publishDone(id: Int) {
        val obj = JSONObject()
        obj.put("id", id)
        obj.put("type", "done")
        emit(obj)
    }

    fun publishError(id: Int, message: String) {
        val obj = JSONObject()
        obj.put("id", id)
        obj.put("type", "error")
        obj.put("message", message)
        emit(obj)
    }

    private fun emit(payload: JSONObject) {
        val json = payload.toString()
        for ((_, fn) in listeners) fn(json)
    }
}

File: packages/lynx-http/ios/HttpEventBus.swift
import Foundation

/// Process-wide pub/sub bus that owns the conversion of native HTTP
/// delegate callbacks into JS-side event payloads. URLSession delegates
/// write here; per-LynxView `HttpPublisher` instances read.
///
/// Mirrors `WebSocketEventBus` — native lifecycle is decoupled from any
/// specific LynxView so in-flight requests survive view recreation.
///
/// Payload keys match the `NativeHttpEvent` interface in `src/types.ts`.
/// Event order per request id: response (once) → progress* → chunk* →
/// done | error.
final class HttpEventBus {

    static let shared = HttpEventBus()

    typealias Payload = [String: Any]
    /// Listeners receive each event as a **JSON string**, not a structured
    /// map — see `emit(_:)`.
    typealias Listener = (String) -> Void

    private let queue = DispatchQueue(label: "com.sigx.http.bus")
    private var listeners: [(token: UUID, fn: Listener)] = []

    @discardableResult
    func addListener(_ fn: @escaping Listener) -> UUID {
        let token = UUID()
        queue.sync { listeners.append((token, fn)) }
        return token
    }

    func removeListener(_ token: UUID) {
        queue.sync { listeners.removeAll { $0.token == token } }
    }

    // MARK: - Publish

    func publish(response status: Int, statusText: String, headers: [String: String], id: Int) {
        emit([
            "id": id,
            "type": "response",
            "status": status,
            "statusText": statusText,
            "headers": headers,
        ])
    }

    func publish(chunk base64: String, id: Int) {
        emit([
            "id": id,
            "type": "chunk",
            "data": base64,
        ])
    }

    func publish(progress loaded: Int64, total: Int64, id: Int) {
        emit([
            "id": id,
            "type": "progress",
            "loaded": loaded,
            "total": total,
        ])
    }

    func publish(done id: Int) {
        emit([
            "id": id,
            "type": "done",
        ])
    }

    func publish(error message: String, id: Int) {
        emit([
            "id": id,
            "type": "error",
            "message": message,
        ])
    }

    private func emit(_ payload: Payload) {
        // Lynx 0.5.0 / PrimJS 3.8 regressed `sendGlobalEvent` marshalling of
        // structured params: the one payload carrying a nested map (the
        // `response` event's `headers`) loses its sibling scalar fields
        // (`status`/`statusText`) crossing the bridge, so `Response.status`
        // arrives undefined and `res.ok` is false for every request (#342).
        // Strings cross intact (the base64 body chunks already do), so emit a
        // single JSON string — the JS shim already parses string-form events.
        guard let data = try? JSONSerialization.data(withJSONObject: payload),
              let json = String(data: data, encoding: .utf8) else { return }
        let snapshot = queue.sync { listeners }
        for (_, fn) in snapshot { fn(json) }
    }
}
